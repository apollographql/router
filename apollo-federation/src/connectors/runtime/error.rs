use serde_json_bytes::{ByteString, Map, Value};

#[derive(Debug, Default)]
pub struct RawError {
    pub message: String,
    pub code: Option<ByteString>,
    pub extensions: Map<ByteString, Value>,
    pub path: Option<ByteString>,
}

impl RawError {
    pub fn extension_code(mut self, code: impl Into<String>) -> Self {
        self.code = Some(code.into().into());
        self
    }

    pub fn extension<K, V>(mut self, key: K, value: V) -> Self
    where
        K: Into<ByteString>,
        V: Into<Value>,
    {
        self.extensions.insert(key.into(), value.into());
        self
    }
}
