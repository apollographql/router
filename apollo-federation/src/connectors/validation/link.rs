//! Validations for `@link(url: "https://specs.apollo.dev/connect")`

use std::fmt::Display;

use apollo_compiler::Name;
use apollo_compiler::Schema;
use apollo_compiler::schema::Component;
use apollo_compiler::schema::Directive;
use itertools::Itertools;
use strum::IntoEnumIterator;

use crate::connectors::ConnectSpec;
use crate::connectors::validation::Code;
use crate::connectors::validation::Message;
use crate::link::Link;

/// The `@link` in a subgraph which enables connectors
#[derive(Clone, Debug)]
pub(crate) struct ConnectLink<'schema> {
    spec: ConnectSpec,
    source_directive_name: Name,
    connect_directive_name: Name,
    directive: &'schema Component<Directive>,
    link: Link,
}

impl<'schema> ConnectLink<'schema> {
    /// Find the connect link, if any, and validate it.
    /// Returns `None` if this is not a connectors subgraph.
    ///
    /// # Errors
    /// - Unknown spec version
    pub(super) fn new(schema: &'schema Schema) -> Option<Result<Self, Message>> {
        let connect_identity = ConnectSpec::identity();
        let (link, directive) = Link::for_identity(schema, &connect_identity)?;

        let spec = match ConnectSpec::try_from(&link.url.version) {
            Err(err) => {
                let message = format!(
                    "{err}; should be one of {available_versions}.",
                    available_versions = ConnectSpec::iter().map(ConnectSpec::as_str).join(", "),
                );
                return Some(Err(Message {
                    code: Code::UnknownConnectorsVersion,
                    message,
                    locations: directive
                        .line_column_range(&schema.sources)
                        .into_iter()
                        .collect(),
                }));
            }
            Ok(spec) => spec,
        };
        let source_directive_name = ConnectSpec::source_directive_name(&link);
        let connect_directive_name = ConnectSpec::connect_directive_name(&link);
        Some(Ok(Self {
            spec,
            source_directive_name,
            connect_directive_name,
            directive,
            link,
        }))
    }

    pub(super) fn spec(&self) -> ConnectSpec {
        self.spec
    }

    pub(super) fn source_directive_name(&self) -> &Name {
        &self.source_directive_name
    }

    pub(super) fn connect_directive_name(&self) -> &Name {
        &self.connect_directive_name
    }

    pub(super) fn directive(&self) -> &Component<Directive> {
        self.directive
    }
}

impl Display for ConnectLink<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.link)
    }
}
