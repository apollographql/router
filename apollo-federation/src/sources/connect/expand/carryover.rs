use apollo_compiler::ast::Argument;
use apollo_compiler::ast::Directive;
use apollo_compiler::ast::Value;
use apollo_compiler::name;
use apollo_compiler::Name;
use apollo_compiler::Node;

use crate::error::FederationError;
use crate::link::inaccessible_spec_definition::INACCESSIBLE_DIRECTIVE_NAME_IN_SPEC;
use crate::link::spec::Identity;
use crate::link::spec::APOLLO_SPEC_DOMAIN;
use crate::link::Link;
use crate::link::DEFAULT_LINK_NAME;
use crate::schema::position::DirectiveArgumentDefinitionPosition;
use crate::schema::position::DirectiveDefinitionPosition;
use crate::schema::position::EnumTypeDefinitionPosition;
use crate::schema::position::EnumValueDefinitionPosition;
use crate::schema::position::InputObjectFieldDefinitionPosition;
use crate::schema::position::InputObjectTypeDefinitionPosition;
use crate::schema::position::InterfaceFieldArgumentDefinitionPosition;
use crate::schema::position::InterfaceFieldDefinitionPosition;
use crate::schema::position::InterfaceTypeDefinitionPosition;
use crate::schema::position::ObjectFieldArgumentDefinitionPosition;
use crate::schema::position::ObjectFieldDefinitionPosition;
use crate::schema::position::ObjectTypeDefinitionPosition;
use crate::schema::position::ScalarTypeDefinitionPosition;
use crate::schema::position::SchemaDefinitionPosition;
use crate::schema::position::UnionTypeDefinitionPosition;
use crate::schema::referencer::DirectiveReferencers;
use crate::schema::FederationSchema;
use crate::sources::connect::spec::LATEST_CONNECT_VERSION;

const TAG_DIRECTIVE_NAME_IN_SPEC: Name = name!("tag");
const AUTHENTICATED_DIRECTIVE_NAME_IN_SPEC: Name = name!("authenticated");
const REQUIRES_SCOPES_DIRECTIVE_NAME_IN_SPEC: Name = name!("requiresScopes");
const POLICY_DIRECTIVE_NAME_IN_SPEC: Name = name!("policy");

pub(super) fn carryover_directives(
    from: &FederationSchema,
    to: &mut FederationSchema,
) -> Result<(), FederationError> {
    let Some(metadata) = from.metadata() else {
        return Ok(());
    };

    // @join__directive(graph: [], name: "link", args: { url: "https://specs.apollo.dev/connect/v0.1" })
    // this must exist for license key enforcement
    SchemaDefinitionPosition.insert_directive(
        to,
        LATEST_CONNECT_VERSION.join_directive_application().into(),
    )?;

    // @inaccessible

    if let Some(link) = metadata.for_identity(&Identity::inaccessible_identity()) {
        let directive_name = link.directive_name_in_schema(&INACCESSIBLE_DIRECTIVE_NAME_IN_SPEC);
        from.referencers()
            .get_directive(&directive_name)
            .and_then(|referencers| {
                // because the merge code handles inaccessible, we have to check if the
                // @link and directive definition are already present in the schema
                if referencers.len() > 0
                    && to
                        .metadata()
                        .and_then(|m| m.by_identity.get(&Identity::inaccessible_identity()))
                        .is_none()
                {
                    SchemaDefinitionPosition
                        .insert_directive(to, link.to_directive_application().into())?;
                    copy_directive_definition(from, to, directive_name.clone())?;
                }
                referencers.copy_directives(from, to, &directive_name)
            })?;
    }

    // @tag

    if let Some(link) = metadata.for_identity(&Identity {
        domain: APOLLO_SPEC_DOMAIN.to_string(),
        name: TAG_DIRECTIVE_NAME_IN_SPEC,
    }) {
        let directive_name = link.directive_name_in_schema(&TAG_DIRECTIVE_NAME_IN_SPEC);
        from.referencers()
            .get_directive(&directive_name)
            .and_then(|referencers| {
                if referencers.len() > 0 {
                    SchemaDefinitionPosition
                        .insert_directive(to, link.to_directive_application().into())?;
                    copy_directive_definition(from, to, directive_name.clone())?;
                }
                referencers.copy_directives(from, to, &directive_name)
            })?;
    }

    // @authenticated

    if let Some(link) = metadata.for_identity(&Identity {
        domain: APOLLO_SPEC_DOMAIN.to_string(),
        name: AUTHENTICATED_DIRECTIVE_NAME_IN_SPEC,
    }) {
        let directive_name = link.directive_name_in_schema(&AUTHENTICATED_DIRECTIVE_NAME_IN_SPEC);
        from.referencers()
            .get_directive(&directive_name)
            .and_then(|referencers| {
                if referencers.len() > 0 {
                    SchemaDefinitionPosition
                        .insert_directive(to, link.to_directive_application().into())?;
                    copy_directive_definition(from, to, directive_name.clone())?;
                }
                referencers.copy_directives(from, to, &directive_name)
            })?;
    }

    // @requiresScopes

    if let Some(link) = metadata.for_identity(&Identity {
        domain: APOLLO_SPEC_DOMAIN.to_string(),
        name: REQUIRES_SCOPES_DIRECTIVE_NAME_IN_SPEC,
    }) {
        let directive_name = link.directive_name_in_schema(&REQUIRES_SCOPES_DIRECTIVE_NAME_IN_SPEC);
        from.referencers()
            .get_directive(&directive_name)
            .and_then(|referencers| {
                if referencers.len() > 0 {
                    SchemaDefinitionPosition
                        .insert_directive(to, link.to_directive_application().into())?;

                    let scalar_type_pos = ScalarTypeDefinitionPosition {
                        type_name: link.type_name_in_schema(&name!(Scope)),
                    };
                    scalar_type_pos
                        .get(from.schema())
                        .map_err(From::from)
                        .and_then(|def| {
                            scalar_type_pos.pre_insert(to)?;
                            scalar_type_pos.insert(to, def.clone())
                        })?;

                    copy_directive_definition(from, to, directive_name.clone())?;
                }
                referencers.copy_directives(from, to, &directive_name)
            })?;
    }

    // @policy

    if let Some(link) = metadata.for_identity(&Identity {
        domain: APOLLO_SPEC_DOMAIN.to_string(),
        name: POLICY_DIRECTIVE_NAME_IN_SPEC,
    }) {
        let directive_name = link.directive_name_in_schema(&POLICY_DIRECTIVE_NAME_IN_SPEC);
        from.referencers()
            .get_directive(&directive_name)
            .and_then(|referencers| {
                if referencers.len() > 0 {
                    SchemaDefinitionPosition
                        .insert_directive(to, link.to_directive_application().into())?;

                    let scalar_type_pos = ScalarTypeDefinitionPosition {
                        type_name: link.type_name_in_schema(&name!(Policy)),
                    };
                    scalar_type_pos
                        .get(from.schema())
                        .map_err(From::from)
                        .and_then(|def| {
                            scalar_type_pos.pre_insert(to)?;
                            scalar_type_pos.insert(to, def.clone())
                        })?;

                    copy_directive_definition(from, to, directive_name.clone())?;
                }
                referencers.copy_directives(from, to, &directive_name)
            })?;
    }

    // compose directive

    metadata
        .directives_by_imported_name
        .iter()
        .filter(|(_name, (link, _import))| !is_known_link(link))
        .try_for_each(|(name, (link, _import))| {
            let directive_name = link.directive_name_in_schema(name);
            from.referencers()
                .get_directive(&directive_name)
                .and_then(|referencers| {
                    if referencers.len() > 0 {
                        if !SchemaDefinitionPosition
                            .get(to.schema())
                            .directives
                            .iter()
                            .any(|d| {
                                d.name == DEFAULT_LINK_NAME
                                    && d.specified_argument_by_name("url")
                                        .and_then(|url| url.as_str())
                                        .map(|url| link.url.to_string() == *url)
                                        .unwrap_or_default()
                            })
                        {
                            SchemaDefinitionPosition
                                .insert_directive(to, link.to_directive_application().into())?;
                        }

                        copy_directive_definition(from, to, directive_name.clone())?;
                    }
                    referencers.copy_directives(from, to, &directive_name)
                })?;
            Ok::<_, FederationError>(())
        })?;

    Ok(())
}

fn is_known_link(link: &Link) -> bool {
    link.url.identity.domain == APOLLO_SPEC_DOMAIN
        && [
            name!(link),
            name!(join),
            name!(tag),
            name!(inaccessible),
            name!(authenticated),
            name!(requiresScopes),
            name!(policy),
        ]
        .contains(&link.url.identity.name)
}

fn copy_directive_definition(
    from: &FederationSchema,
    to: &mut FederationSchema,
    directive_name: Name,
) -> Result<(), FederationError> {
    let def_pos = DirectiveDefinitionPosition { directive_name };

    def_pos
        .get(from.schema())
        .map_err(From::from)
        .and_then(|def| {
            def_pos.pre_insert(to)?;
            def_pos.insert(to, def.clone())
        })
}

impl Link {
    fn to_directive_application(&self) -> Directive {
        let mut arguments: Vec<Node<Argument>> = vec![Argument {
            name: name!(url),
            value: self.url.to_string().into(),
        }
        .into()];

        // purpose: link__Purpose
        if let Some(purpose) = &self.purpose {
            arguments.push(
                Argument {
                    name: name!(for),
                    value: Value::Enum(purpose.into()).into(),
                }
                .into(),
            );
        }

        // as: String
        if let Some(alias) = &self.spec_alias {
            arguments.push(
                Argument {
                    name: name!(as),
                    value: Value::String(alias.to_string()).into(),
                }
                .into(),
            );
        }

        // import: [link__Import!]
        if !self.imports.is_empty() {
            arguments.push(
                Argument {
                    name: name!(import),
                    value: Value::List(
                        self.imports
                            .iter()
                            .map(|i| {
                                let name = if i.is_directive {
                                    format!("@{}", i.element)
                                } else {
                                    i.element.to_string()
                                };

                                if let Some(alias) = &i.alias {
                                    let alias = if i.is_directive {
                                        format!("@{}", alias)
                                    } else {
                                        alias.to_string()
                                    };

                                    Value::Object(vec![
                                        (name!(name), Value::String(name).into()),
                                        (name!(alias), Value::String(alias).into()),
                                    ])
                                } else {
                                    Value::String(name)
                                }
                                .into()
                            })
                            .collect::<Vec<_>>(),
                    )
                    .into(),
                }
                .into(),
            );
        }

        Directive {
            name: name!(link),
            arguments,
        }
    }
}

trait CopyDirective {
    fn copy_directive(
        &self,
        from: &FederationSchema,
        to: &mut FederationSchema,
        directive_name: &Name,
    ) -> Result<(), FederationError>;
}

impl CopyDirective for SchemaDefinitionPosition {
    fn copy_directive(
        &self,
        from: &FederationSchema,
        to: &mut FederationSchema,
        directive_name: &Name,
    ) -> Result<(), FederationError> {
        self.get(from.schema())
            .directives
            .iter()
            .filter(|d| &d.name == directive_name)
            .try_for_each(|directive| self.insert_directive(to, directive.clone()))
    }
}

macro_rules! impl_copy_directive {
    ($( $Ty: ty )+) => {
        $(
            impl CopyDirective for $Ty {
                fn copy_directive(
                    &self,
                    from: &FederationSchema,
                    to: &mut FederationSchema,
                    directive_name: &Name,
                ) -> Result<(), FederationError> {
                    self.get(from.schema())
                        .map(|def| {
                            def.directives
                                .iter()
                                .filter(|d| &d.name == directive_name)
                                .try_for_each(|directive| self.insert_directive(to, directive.clone()))
                        })
                        .unwrap_or(Ok(()))
                }
            }
        )+
    };
}

impl_copy_directive! {
    ScalarTypeDefinitionPosition
    ObjectTypeDefinitionPosition
    ObjectFieldDefinitionPosition
    ObjectFieldArgumentDefinitionPosition
    InterfaceTypeDefinitionPosition
    InterfaceFieldDefinitionPosition
    InterfaceFieldArgumentDefinitionPosition
    UnionTypeDefinitionPosition
    EnumTypeDefinitionPosition
    EnumValueDefinitionPosition
    InputObjectTypeDefinitionPosition
    InputObjectFieldDefinitionPosition
    DirectiveArgumentDefinitionPosition
}

impl DirectiveReferencers {
    fn len(&self) -> usize {
        self.schema.as_ref().map(|_| 1).unwrap_or_default()
            + self.scalar_types.len()
            + self.object_types.len()
            + self.object_fields.len()
            + self.object_field_arguments.len()
            + self.interface_types.len()
            + self.interface_fields.len()
            + self.interface_field_arguments.len()
            + self.union_types.len()
            + self.enum_types.len()
            + self.enum_values.len()
            + self.input_object_types.len()
            + self.input_object_fields.len()
            + self.directive_arguments.len()
    }

    fn copy_directives(
        &self,
        from: &FederationSchema,
        to: &mut FederationSchema,
        directive_name: &Name,
    ) -> Result<(), FederationError> {
        if let Some(position) = &self.schema {
            position.copy_directive(from, to, directive_name)?
        }
        self.scalar_types
            .iter()
            .try_for_each(|position| position.copy_directive(from, to, directive_name))?;
        self.object_types
            .iter()
            .try_for_each(|position| position.copy_directive(from, to, directive_name))?;
        self.object_fields
            .iter()
            .try_for_each(|position| position.copy_directive(from, to, directive_name))?;
        self.object_field_arguments
            .iter()
            .try_for_each(|position| position.copy_directive(from, to, directive_name))?;
        self.interface_types
            .iter()
            .try_for_each(|position| position.copy_directive(from, to, directive_name))?;
        self.interface_fields
            .iter()
            .try_for_each(|position| position.copy_directive(from, to, directive_name))?;
        self.interface_field_arguments
            .iter()
            .try_for_each(|position| position.copy_directive(from, to, directive_name))?;
        self.union_types
            .iter()
            .try_for_each(|position| position.copy_directive(from, to, directive_name))?;
        self.enum_types
            .iter()
            .try_for_each(|position| position.copy_directive(from, to, directive_name))?;
        self.enum_values
            .iter()
            .try_for_each(|position| position.copy_directive(from, to, directive_name))?;
        self.input_object_types
            .iter()
            .try_for_each(|position| position.copy_directive(from, to, directive_name))?;
        self.input_object_fields
            .iter()
            .try_for_each(|position| position.copy_directive(from, to, directive_name))?;
        self.directive_arguments
            .iter()
            .try_for_each(|position| position.copy_directive(from, to, directive_name))?;
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use apollo_compiler::Schema;
    use insta::assert_snapshot;

    use super::carryover_directives;
    use crate::merge::merge_federation_subgraphs;
    use crate::schema::FederationSchema;
    use crate::supergraph::extract_subgraphs_from_supergraph;

    #[test]
    fn test_carryover() {
        let sdl = include_str!("./tests/schemas/ignore/directives.graphql");
        let schema = Schema::parse(sdl, "directives.graphql").expect("parse failed");
        let supergraph_schema = FederationSchema::new(schema).expect("federation schema failed");
        let subgraphs = extract_subgraphs_from_supergraph(&supergraph_schema, None)
            .expect("extract subgraphs failed");
        let merged = merge_federation_subgraphs(subgraphs).expect("merge failed");
        let schema = merged.schema.into_inner();
        let mut schema = FederationSchema::new(schema).expect("federation schema failed");

        carryover_directives(&supergraph_schema, &mut schema).expect("carryover failed");
        assert_snapshot!(schema.schema().serialize().to_string());
    }
}
