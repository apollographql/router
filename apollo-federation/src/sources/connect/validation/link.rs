//! Validations for `@link(url: "https://specs.apollo.dev/connect")`

use std::fmt::Display;

use apollo_compiler::Name;
use apollo_compiler::Schema;
use apollo_compiler::schema::Component;
use apollo_compiler::schema::Directive;
use itertools::Itertools;

use crate::link::Link;
use crate::sources::connect::ConnectSpec;
use crate::sources::connect::validation::Code;
use crate::sources::connect::validation::Message;

/// The `@link` in a subgraph which enables connectors
#[derive(Clone, Debug)]
pub(super) struct ConnectLink<'schema> {
    pub(crate) spec: ConnectSpec,
    pub(crate) source_directive_name: Name,
    pub(crate) connect_directive_name: Name,
    pub(crate) directive: &'schema Component<Directive>,
    link: Link,
}

impl<'schema> ConnectLink<'schema> {
    /// Find the connect link, if any, and validate it.
    /// Returns `None` if this is not a connectors subgraph.
    ///
    /// # Errors
    /// - Unknown spec version
    pub(super) fn new(schema: &'schema Schema) -> Option<Result<Self, Message>> {
        let connect_identity = ConnectSpec::identity();
        let (link, directive) = Link::for_identity(schema, &connect_identity)?;

        let spec = match ConnectSpec::try_from(&link.url.version) {
            Err(err) => {
                let available_versions = ConnectSpec::available();
                let message = if available_versions.len() == 1 {
                    // TODO: No need to branch here once multiple spec versions are available
                    format!("{err}; should be {version}.", version = ConnectSpec::V0_1)
                } else {
                    // This won't happen today, but it's prepping for 0.2 so we don't forget
                    format!(
                        "{err}; should be one of {available_versions}.",
                        available_versions = available_versions
                            .iter()
                            .copied()
                            .map(ConnectSpec::as_str)
                            .join(", "),
                    )
                };
                return Some(Err(Message {
                    code: Code::UnknownConnectorsVersion,
                    message,
                    locations: directive
                        .line_column_range(&schema.sources)
                        .into_iter()
                        .collect(),
                }));
            }
            Ok(spec) => spec,
        };
        let source_directive_name = ConnectSpec::source_directive_name(&link);
        let connect_directive_name = ConnectSpec::connect_directive_name(&link);
        Some(Ok(Self {
            spec,
            source_directive_name,
            connect_directive_name,
            directive,
            link,
        }))
    }
}

impl Display for ConnectLink<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "@link(url: \"{}\")", self.link.url)
    }
}
