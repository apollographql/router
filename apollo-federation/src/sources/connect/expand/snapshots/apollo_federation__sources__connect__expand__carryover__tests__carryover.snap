---
source: apollo-federation/src/sources/connect/expand/carryover.rs
expression: schema.schema().serialize().to_string()
---
schema @link(url: "https://specs.apollo.dev/link/v1.0") @link(url: "https://specs.apollo.dev/join/v0.3", for: EXECUTION) @join__directive(graphs: [], name: "link", args: {url: "https://specs.apollo.dev/connect/v0.1"}) @link(url: "https://specs.apollo.dev/inaccessible/v0.2", for: SECURITY) @link(url: "https://specs.apollo.dev/tag/v0.3") @link(url: "https://specs.apollo.dev/authenticated/v0.1", for: SECURITY) @link(url: "https://specs.apollo.dev/requiresScopes/v0.1", for: SECURITY) @link(url: "https://specs.apollo.dev/policy/v0.1", for: SECURITY) @link(url: "http://specs.example.org/custom/v0.1", import: ["@custom1", "@custom2", "@custom3"]) {
  query: Query
}

directive @link(url: String, as: String, for: link__Purpose, import: [link__Import]) repeatable on SCHEMA

directive @join__graph(name: String!, url: String!) on ENUM_VALUE

directive @join__type(graph: join__Graph!, key: join__FieldSet, extension: Boolean! = false, resolvable: Boolean! = true, isInterfaceObject: Boolean! = false) repeatable on ENUM | INPUT_OBJECT | INTERFACE | OBJECT | SCALAR | UNION

directive @join__field(graph: join__Graph, requires: join__FieldSet, provides: join__FieldSet, type: String, external: Boolean, override: String, overrideLabel: String, usedOverridden: Boolean) repeatable on FIELD_DEFINITION | INPUT_FIELD_DEFINITION

directive @join__implements(graph: join__Graph!, interface: String!) repeatable on INTERFACE | OBJECT

directive @join__unionMember(graph: join__Graph!, member: String!) repeatable on UNION

directive @join__enumValue(graph: join__Graph!) repeatable on ENUM_VALUE

directive @join__directive(graphs: [join__Graph!], name: String!, args: join__DirectiveArguments!) repeatable on SCHEMA | OBJECT | INTERFACE | FIELD_DEFINITION

directive @inaccessible on FIELD_DEFINITION | OBJECT | INTERFACE | UNION | ARGUMENT_DEFINITION | SCALAR | ENUM | ENUM_VALUE | INPUT_OBJECT | INPUT_FIELD_DEFINITION

directive @tag(name: String!) repeatable on FIELD_DEFINITION | OBJECT | INTERFACE | UNION | ARGUMENT_DEFINITION | SCALAR | ENUM | ENUM_VALUE | INPUT_OBJECT | INPUT_FIELD_DEFINITION | SCHEMA

directive @authenticated on FIELD_DEFINITION | OBJECT | INTERFACE | SCALAR | ENUM

directive @requiresScopes(scopes: [[requiresScopes__Scope!]!]!) on FIELD_DEFINITION | OBJECT | INTERFACE | SCALAR | ENUM

directive @policy(policies: [[policy__Policy!]!]!) on FIELD_DEFINITION | OBJECT | INTERFACE | SCALAR | ENUM

directive @custom1 on OBJECT | FIELD_DEFINITION

directive @custom2 on OBJECT | FIELD_DEFINITION

directive @custom3 on OBJECT | FIELD_DEFINITION

enum link__Purpose {
  """
  SECURITY features provide metadata necessary to securely resolve fields.
  """
  SECURITY
  """EXECUTION features provide metadata necessary for operation execution."""
  EXECUTION
}

scalar link__Import

scalar join__FieldSet

scalar join__DirectiveArguments

enum join__Graph {
  ONE @join__graph(name: "one", url: "none")
  TWO @join__graph(name: "two", url: "none")
}

type Query @join__type(graph: ONE) @join__type(graph: TWO) {
  tagged: String @join__field(graph: ONE) @tag(name: "tag")
  hidden: String @join__field(graph: ONE) @inaccessible
  custom: T @join__field(graph: ONE) @custom1
  authenticated: String @join__field(graph: ONE) @authenticated
  requiresScopes: String @join__field(graph: ONE) @requiresScopes(scopes: ["scope"])
  policy: String @join__field(graph: ONE) @policy(policies: [["admin"]])
  overridden: String @join__field(graph: ONE, override: "two", overrideLabel: "label") @join__field(graph: TWO)
  customAgain: String @join__field(graph: TWO) @custom1
}

type T @join__type(graph: ONE) @custom2 {
  field: String @join__field(graph: ONE) @custom3
}

scalar requiresScopes__Scope

scalar policy__Policy
